base/sparse/higherorderfns.jl

```diff
@@ -986,41 +986,97 @@ broadcast{Tf,T}(f::Tf, ::Type{T}, A::SparseMatrixCSC) = broadcast(y -> f(T, y),
-promote_containertype(::Type{StructuredArray}, ::Type{Array}) = Array
-promote_containertype(::Type{Array}, ::Type{StructuredArray}) = Array
-# combinations involving tuples divert to the generic array code
-promote_containertype(::Type{StructuredArray}, ::Type{Tuple}) = Array
-promote_containertype(::Type{Tuple}, ::Type{StructuredArray}) = Array
 
-# for combinations involving sparse/structured arrays and scalars only,
-# promote all structured arguments to sparse and then rebroadcast
-@inline broadcast_c{N}(f, ::Type{StructuredArray}, As::Vararg{Any,N}) =
+# first (Broadcast containertype) dispatch layer's promotion logic
+immutable PromoteToSparse end
+
+# broadcast containertype definitions for structured matrices
+StructuredMatrix = Union{Diagonal,Bidiagonal,Tridiagonal,SymTridiagonal}
+_containertype{T<:StructuredMatrix}(::Type{T}) = PromoteToSparse
+
+# combinations explicitly involving Tuples and PromoteToSparse collections
+# divert to the generic AbstractArray broadcast code
+promote_containertype(::Type{PromoteToSparse}, ::Type{Tuple}) = Array
+promote_containertype(::Type{Tuple}, ::Type{PromoteToSparse}) = Array
+# combinations involving scalars and PromoteToSparse collections continue in the promote-to-sparse funnel
+promote_containertype(::Type{PromoteToSparse}, ::Type{Any}) = PromoteToSparse
+promote_containertype(::Type{Any}, ::Type{PromoteToSparse}) = PromoteToSparse
+# combinations involving sparse arrays and PromoteToSparse collections continue in the promote-to-sparse funnel
+promote_containertype(::Type{PromoteToSparse}, ::Type{AbstractSparseArray}) = PromoteToSparse
+promote_containertype(::Type{AbstractSparseArray}, ::Type{PromoteToSparse}) = PromoteToSparse
+# combinations involving Arrays and PromoteToSparse ecollections continue in the promote-to-sparse funnel
+promote_containertype(::Type{PromoteToSparse}, ::Type{Array}) = PromoteToSparse
+promote_containertype(::Type{Array}, ::Type{PromoteToSparse}) = PromoteToSparse
+# combinations involving Arrays and sparse arrays continue in the promote-to-sparse funnel
+promote_containertype(::Type{AbstractSparseArray}, ::Type{Array}) = PromoteToSparse
+promote_containertype(::Type{Array}, ::Type{AbstractSparseArray}) = PromoteToSparse
+
+# second (internal sparse broadcast containertype) dispatch layer's promotion logic
+# mostly just disambiguates Array from the main containertype promotion mechanism
+# AbstractArray serves as a marker to shunt to the generic AbstractArray broadcast code
+_spcontainertype(x) = _containertype(x)
+_spcontainertype{T<:Vector}(::Type{T}) = Vector
+_spcontainertype{T<:Matrix}(::Type{T}) = Matrix
+_spcontainertype{T<:Ref}(::Type{T}) = AbstractArray
+_spcontainertype{T<:AbstractArray}(::Type{T}) = AbstractArray
+# need the following two methods to override the immediately preceding method
+_spcontainertype{T<:StructuredMatrix}(::Type{T}) = PromoteToSparse
+_spcontainertype{T<:SparseVecOrMat}(::Type{T}) = AbstractSparseArray
```

Comment:
We should think through how the AbstractSparse types are used a bit, maybe not right here right now though. Anything that's assuming the exact CSC or concrete SparseVector data structures, fields, storage order etc should only be defined for those concrete types. But any other behavior we think should generically apply for all sparse arrays, I think it's worth widening some signatures and letting the definers of any concrete subtypes override more specific behaviors where they want.

I guess in this particular case the tradeoff would be between getting method errors vs dispatching to more generic dense behavior? The latter's likely better for now until we think through a better extensibility mechanism for how to get broadcast on custom array types to do interesting things.